### PR TITLE
Service with type=NodePort and clusterIP=None is not allowed

### PR DESCRIPTION
**Pull request by derekwaynecarr**
Fixes https://github.com/kubernetes/kubernetes/issues/30908

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30932)

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
cc @thockin ask and ye shall receive (was looking at another validation issue, and saw this was easy)

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -2189,6 +2189,9 @@ func ValidateService(service *api.Service) field.ErrorList {
 		allErrs = append(allErrs, field.Required(specPath.Child("ports"), ""))
 	}
 	if service.Spec.Type == api.ServiceTypeLoadBalancer {
+		if service.Spec.ClusterIP == "None" {
+			allErrs = append(allErrs, field.Invalid(specPath.Child("clusterIP"), service.Spec.ClusterIP, "invalid value when using load balancer service type"))
```
Comment (thockin) [inline pkg/api/validation/validation.go]: can you also check for type == NodePort?  I opened 2 bugs, but it is the same problem.

**Comment by thockin** [comment]
Thanks.

This should probably have a test case, too :)

**Comment by thockin** [comment]
Derek, do you need someone to take over this?

**Comment by derekwaynecarr** [comment]
I will rebase this today, sorry for delay.

**Comment by derekwaynecarr** [comment]
@thockin -- the original issue was fixed by another PR, I updated this PR to fix https://github.com/kubernetes/kubernetes/issues/30908  

sorry for the delay on this one.

File: pkg/api/validation/validation_test.go [hunk 2]
```diff
@@ -5207,6 +5207,15 @@ func TestValidateService(t *testing.T) {
 			},
 			numErrs: 1,
 		},
+		{
+			name: "invalid node port with clusterIP None",
```
Comment (thockin) [inline pkg/api/validation/validation_test.go]: can you repeat this case in  TestValidateServiceUpdate

**Comment by derekwaynecarr** [comment]
@thockin - updated with test on update as requested.

**Comment by derekwaynecarr** [comment]
@k8s-bot gci gke e2e test this

**Comment by derekwaynecarr** [comment]
Review status: 0 of 2 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/api/validation/validation.go, line 2193 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30932#-KPUcjT3ddw5krJE4evp-r1-2193:-KYytKaCoZOJ8u1KGsO3:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/8053f0db1fe3623c35b58d76dd76fdb0b1c31014/pkg/api/validation/validation.go#L2193)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary><blockquote>

can you also check for type == NodePort?  I opened 2 bugs, but it is the same problem.
</blockquote></details>

Done.

---

*[pkg/api/validation/validation_test.go, line 5211 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30932#-KPUcjT3ddw5krJE4evq-r2-5211:-KYytFdNKCOqrwMbWeVn:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/68985b457094de7bf5ae75be918fef75fdd49acb/pkg/api/validation/validation_test.go#L5211)):*
<details><summary><i>Previously, thockin (Tim Hockin) wrote…</i></summary><blockquote>

can you repeat this case in  TestValidateServiceUpdate
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30932)*

**Comment by thockin** [comment]
@k8s-bot cvm gke e2e test this

**Pull request merged**